app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -4,6 +4,16 @@ class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
   # You should configure your model like this:
   # devise :omniauthable, omniauth_providers: [:twitter]
 
+  def google_oauth2
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+    if @user.persisted?
+      sign_in @user
+      redirect_to user_path(@user.id), notice: 'Successfully logged in using your Google account.'
+    else
+      session["devise.google_data"] = request.env["omniauth.auth"]
+      redirect_to root_path, alert: 'Something went wrong. Please try to register instead.'
```

Comment:
Can we go over what devise.google_data is doing?

Reply:
I finally found the resource I used to implement this controller. Here's the explanation regarding what session["devise.google_data"] does:
```In case the user is not persisted, we store the OmniAuth data in the session. Notice we store this data using "devise." as key namespace. This is useful because Devise removes all the data starting with "devise." from the session whenever a user signs in, so we get automatic session clean up. At the end, we redirect the user back to our registration form.```
Also, here's the link with more info for future reference: https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview

Reply:
Strong work! Good job working with wha we have to implement OAuth. Interesting to see how this is done in tandem with devise.